### PR TITLE
fix: Avoid crashing in NativeViewHost::SetParentAccessible on Windows 10

### LINKED ISSUE #26905: Windows crash in views::NativeViewHost::SetParentAccessible
We're getting lots of crash reports like this on Electron 10.1.6.

Here's one stack (this can be avoided by `--disable-renderer-accessibility`):

```
Fatal Error: EXCEPTION_ACCESS_VIOLATION_READ

Thread 19604 Crashed:
0   Figma.exe                       0x7ff7fa80d457      views::NativeViewHost::SetParentAccessible (native_view_host.cc:57)
1   Figma.exe                       0x7ff7f791f997      [inlined] base::ObserverList<T>::Iter::EnsureValidIndex (observer_list.h:214)
2   Figma.exe                       0x7ff7f791f997      [inlined] base::ObserverList<T>::Iter::Iter (observer_list.h:136)
3   Figma.exe                       0x7ff7f791f997      [inlined] base::ObserverList<T>::begin (observer_list.h:240)
4   Figma.exe                       0x7ff7f791f997      ui::AXPlatformNode::NotifyAddAXModeFlags (ax_platform_node.cc:98)
5   Figma.exe                       0x7ff7f88046cd      content::BrowserAccessibilityStateImpl::AddAccessibilityModeFlags (browser_accessibility_state_impl.cc:224)
6   Figma.exe                       0x7ff7f795b77c      ui::AXPlatformNodeWin::QueryService (ax_platform_node_win.cc:4531)
```

For full stack trace, see: https://sentry.io/share/issue/9d0dab36a8c84cdb84cb6bde1ed39b32/

Here's another (this happens regardless of `--disable-renderer-accessibility`):

```
0   Figma.exe                       0x7ff681ffd707      views::NativeViewHost::SetParentAccessible (native_view_host.cc:57)
1   Figma.exe                       0x7ff67f10fc17      [inlined] base::ObserverList<T>::Iter::EnsureValidIndex (observer_list.h:214)
2   Figma.exe                       0x7ff67f10fc17      [inlined] base::ObserverList<T>::Iter::Iter (observer_list.h:136)
3   Figma.exe                       0x7ff67f10fc17      [inlined] base::ObserverList<T>::begin (observer_list.h:240)
4   Figma.exe                       0x7ff67f10fc17      ui::AXPlatformNode::NotifyAddAXModeFlags (ax_platform_node.cc:98)
5   Figma.exe                       0x7ff67f13f16b      ui::AXPlatformNodeWin::get_attributes (ax_platform_node_win.cc:1363)
6   UIAutomationCore.DLL            0x7ffbe4bf8bc1      AccUtils::get_attributes
7   UIAutomationCore.DLL            0x7ffbe4c5a034      MsaaProxy::GetLiveSetting
8   UIAutomationCore.DLL            0x7ffbe4c5977f      MsaaProxy::GetIsControlElementPropertyValue
9   UIAutomationCore.DLL            0x7ffbe4b75253      destructor'
10  UIAutomationCore.DLL            0x7ffbe4b02e0e      MsaaProxy::GetPropertyValue
11  UIAutomationCore.DLL            0x7ffbe4b27054      UiaNode::ProviderGetPropertyValue
12  UIAutomationCore.DLL            0x7ffbe4b26b7b      InProcClientAPIStub::UiaNode_GetPropertyValues
13  UIAutomationCore.DLL            0x7ffbe4b1947d      ComInvoker::CallTarget
14  UIAutomationCore.DLL            0x7ffbe4b2c228      InProcClientAPIStub::InvokeInProcAPI
15  UIAutomationCore.DLL            0x7ffbe4b407f7      UiaNode::CrossProcess_GetPropertyValues
16  UIAutomationCore.DLL            0x7ffbe4b0567a      RemoteUiaNodeStub::Incoming_GetPropertyValues
17  UIAutomationCore.DLL            0x7ffbe4b08f7f      RemoteUiaNodeStub::OnMessage
18  UIAutomationCore.DLL            0x7ffbe4b0585b      InvokeOnCorrectContext_Callback
19  UIAutomationCore.DLL            0x7ffbe4b1947d      ComInvoker::CallTarget
20  UIAutomationCore.DLL            0x7ffbe4b204e6      ProcessIncomingRequest
21  UIAutomationCore.DLL            0x7ffbe4b1fc38      HookBasedServerConnectionManager::HookCallback
22  UIAutomationCore.DLL            0x7ffbe4b0a316      HookUtil<T>::CallOut
23  UIAutomationCore.DLL            0x7ffbe4aeacf4      HandleSyncHookMessage
24  UIAutomationCore.DLL            0x7ffbe4aea9b6      HookUtil<T>::CallWndProc
25  USER32.dll                      0x7ffc13c8e021      fnHkINLPCWPSTRUCTW
26  USER32.dll                      0x7ffc13c90bc2      _fnDWORD
27  ntdll.dll                       0x7ffc155cfbd3      KiUserCallbackDispatcher
28  win32u.dll                      0x7ffc13561063      ZwUserPeekMessage
29  USER32.dll                      0x7ffc13c7a5c2      _PeekMessage
30  USER32.dll                      0x7ffc13c7a522      PeekMessageW
31  Figma.exe                       0x7ff680bcf3e5      base::MessagePumpForUI::ProcessNextWindowsMessage (message_pump_win.cc:500)
32  Figma.exe                       0x7ff680bcf2a4      base::MessagePumpForUI::DoRunLoop (message_pump_win.cc:225)
33  Figma.exe                       0x7ff680bcecd4      base::MessagePumpWin::Run (message_pump_win.cc:86)
```

Full trace: https://sentry.io/share/issue/6d1b9c9aa09c457795b3f255b27d6f29/

Here is the OS distribution:

% | version | last seen
-- | -- | --
55.66% | Windows 10.0.18362 | a minute ago
44.26% | Windows 10.0.19041 | 2 minutes ago
0.03% | Windows 10.0.17763 | 11 hours ago
0.02% | Windows 10.0.20211 | 4 days ago
0.02% | Windows 10.0.20221 | 3 days ago
0.02% | Windows 10.0.20270 | 18 hours ago

### PR DESCRIPTION
**Pull request by poiru**
This fixes #26905. The patch was obtained from @deepak1556, who in turn
got it from the Microsoft Teams folks.

I believe the crash started happening due to the changes in
https://chromium.googlesource.com/chromium/src.git/+/5c6c8e994bce2bfb867279ae5068e9f9134e70c3%5E!/#F15

This affects Electron 9 and later.

Notes: Fix occasional crash on Windows

### CODE REVIEW

File: patches/chromium/fix_setparentacessibile_crash_win.patch [hunk 1]
```diff
@@ -0,0 +1,25 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Biru Mohanathas <birunthan@mohanathas.com>
+Date: Thu, 10 Dec 2020 19:02:37 +0200
+Subject: fix crash in NativeViewHost::SetParentAccessible
+
+This fixes random crashes on Windows 10. It presumably started happening
+after the changes in
+https://chromium.googlesource.com/chromium/src.git/+/5c6c8e994bce2bfb867279ae5068e9f9134e70c3%5E!/#F15
+
+For context, see: https://github.com/electron/electron/issues/26905
```
Comment (nornagon) [inline patches/chromium/fix_setparentacessibile_crash_win.patch:10]: We require all patches to describe a plan for their own removal, or else a reason why they should be maintained forever, per https://github.com/electron/electron/blob/master/docs/development/patches.md#patch-justification.

Can this patch be upstreamed? If not, why not? Can we fix this crash without a patch?

**Review by nornagon — CHANGES_REQUESTED** [review]

File: patches/chromium/fix_setparentacessibile_crash_win.patch (hunk 1 above)
Comment (poiru) [reply patches/chromium/fix_setparentacessibile_crash_win.patch:10]: @nornagon Updated!

**Review by miniak — APPROVED** [review]

**Comment by deepak1556** [comment]
This is based on the investigation done by Way Vadhanasin and Julie Koubova from MS teams. The original trace from memory dump was

```
views::NativeViewHost::SetParentAccessible(IAccessible * accessible) Line 57 C++
ui::AXPlatformNode::NotifyAddAXModeFlags(ui::AXMode mode_flags) Line 97 C++
content::BrowserAccessibilityStateImpl::AddAccessibilityModeFlags(ui::AXMode mode) Line 227 C++
ui::AXPlatformNodeWin::QueryService(const _GUID & guidService, const _GUID & riid, void * * object) Line 4529 C++
oleacc.dll!AccWrap_Base::QueryService() Unknown
nvdaHelperRemote.dll!00007ffd31ff8dff() Unknown
nvdaHelperRemote.dll!00007ffd31fec10e() Unknown
nvdaHelperRemote.dll!00007ffd31fe3afb() Unknown
user32.dll!__ClientCallWinEventProc() Unknown
ntdll.dll!KiUserCallbackDispatcherContinue() Unknown
win32u.dll!NtUserNotifyWinEvent() Unknown
user32.dll!NotifyWinEvent() Unknown
ui::AXSystemCaretWin::MoveCaretTo(const gfx::Rect & bounds_physical_pixels) Line 72 C++
views::HWNDMessageHandler::OnCaretBoundsChanged(const ui::TextInputClient * client) Line 1060 C++
ui::InputMethodBase::NotifyTextInputCaretBoundsChanged(const ui::TextInputClient * client) Line 158 C++
ui::InputMethodWinTSF::OnCaretBoundsChanged(const ui::TextInputClient * client) Line 111 C++
ui::InputMethodWinTSF::OnDidChangeFocusedClient(ui::TextInputClient * focused_before, ui::TextInputClient * focused) Line 168 C++
[Inline Frame] ui::InputMethodBase::SetFocusedTextInputClientInternal(ui::TextInputClient * client) Line 169 C++
ui::InputMethodBase::SetFocusedTextInputClient(ui::TextInputClient * client) Line 53 C++
content::RenderWidgetHostViewAura::OnWindowFocused(aura::Window * gained_focus, aura::Window * lost_focus) Line 1800 C++
wm::FocusController::SetFocusedWindow(aura::Window * window) Line 287 C++
wm::FocusController::FocusAndActivateWindow(wm::ActivationChangeObserver::ActivationReason reason, aura::Window * window) Line 244 C++
views::View::Focus() Line 2001 C++
views::FocusManager::SetFocusedViewWithReason(views::View * view, views::FocusManager::FocusChangeReason reason) Line 378 C++
views::FocusManager::RestoreFocusedView() Line 456 C++
views::DesktopNativeWidgetAura::OnWindowActivated(wm::ActivationChangeObserver::ActivationReason reason, aura::Window * gained_active, aura::Window * lost_active) Line 1182 C++
electron::ElectronDesktopNativeWidgetAura::OnWindowActivated(wm::ActivationChangeObserver::ActivationReason reason, aura::Window * gained_active, aura::Window * lost_active) Line 47 C++
wm::FocusController::SetActiveWindow(wm::ActivationChangeObserver::ActivationReason reason, aura::Window * requested_window, aura::Window * window) Line 366 C++
wm::FocusController::FocusAndActivateWindow(wm::ActivationChangeObserver::ActivationReason reason, aura::Window * window) Line 217 C++
electron::api::BrowserWindow::OnWindowFocus() Line 287 C++
```

Inspecting the dump had showed that [NativeViewHost](https://source.chromium.org/chromium/chromium/src/+/master:ui/views/controls/native/native_view_host.h) does not have a valid [native_wrapper_](https://source.chromium.org/chromium/chromium/src/+/master:ui/views/controls/native/native_view_host.h;l=134) set. The instance of NativeViewHost comes from a [holder_](https://source.chromium.org/chromium/chromium/src/+/master:ui/views/controls/webview/webview.h;l=180) member of WebView. And this causes the Ax notification to dereference a null pointer.


Tracing the construction of the above WebView instance

```
views::WebView::WebView
electron::InspectableWebContentsViewViews::InspectableWebContentsViewViews
electron::InspectableWebContentsImpl::InspectableWebContentsImpl
electron::CommonWebContentsDelegate::InitWithWebContents
electron::api::WebContents::InitWithSessionAndOptions
electron::api::WebContents::WebContents
electron::api::WebContents::Create
base::internal::FunctorTraits<v8::Local<v8::Promise> (*)(v8::Isolate *, const base::trace_event::TraceConfig &),void>::Invoke
base::internal::InvokeHelper<0,v8::Local<v8::Promise> >::MakeItSo
base::internal::Invoker<base::internal::BindState<v8::Local<v8::Promise> (*)(…
base::internal::Invoker<base::internal::BindState<v8::Local<v8::Promise> (*)(…
base::RepeatingCallback<gin::Handle<electron::api::WebContents> (v8::Isolate *, const gin_helper::Dictionary &)>::Run
gin_helper::Invoker<gin_helper::IndicesHolder<0,1>,v8::Isolate *,const gin_helper::Dictionary &>::DispatchToCallback
gin_helper::Dispatcher<gin::Handle<electron::api::WebContents> (v8::Isolate *, const gin_helper::Dictionary &)>::DispatchToCallback
v8::internal::FunctionCallbackArguments::Call
```

It comes from https://github.com/electron/electron/blob/b9c9e7fc06772a45ad003fcbe422c114ca65ea5a/shell/browser/ui/views/inspectable_web_contents_view_views.cc#L86-L92

The reason why WebView  [doesn't get a native_wrapper_](https://source.chromium.org/chromium/chromium/src/+/master:ui/views/controls/native/native_view_host.cc;l=180-202) is because the parent of these view's which is electron::InspectableWebContentsViewViews is *not* of type RootView where there is an enforcement of a [valid Widget](https://source.chromium.org/chromium/chromium/src/+/master:ui/views/view.cc;l=239-242)

**Comment by deepak1556** [comment]
I am not sure of the reason why we consider `electron::InspectableWebContentsViewViews` as Root, @zcbenz might more context on this. But I thought https://github.com/electron/electron/pull/22739 might fix the RootView assumption and it did not.

So for now teams have incorporated this patch that fixes the crash path but definitely doesn't address the root issue, which is with how we maintain the view tree.

**Review by nornagon — APPROVED** [review]
Accepted for now, I hope that we can find a fix for this that does not require a patch.

**Pull request merged**